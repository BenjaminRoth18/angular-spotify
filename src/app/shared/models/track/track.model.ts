export interface Track {
    artist: string;
    track: string;
    imageUrl: string;
}

export interface PlaylistDto {
    tracks: {
        items: PlaylistItemDto[];
    };
}

export interface PlaylistItemDto {
    track: PlaylistTrackDto;
}

export interface PlaylistTrackDto {
    album: {
        images: Array<{ url: string }>;
    };
    artists: Array<{ name: string }>;
    name: string;
}
