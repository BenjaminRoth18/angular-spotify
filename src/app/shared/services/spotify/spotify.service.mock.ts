import { SpotifyService } from './spotify.service';

export class MockSpotifyService extends SpotifyService {
    tokenEndpoint = 'tokenEndpoint';
    tokenParams = 'tokenParams';
    playlistEndpoint = 'playlistEndpoint';
    createTokenHeaders = jest.fn().mockReturnValue('tokenHeaders');
    createAuthHeaders = jest.fn().mockReturnValue('authHeaders');
    createPlaylist = jest.fn();

    constructor() {
        super({} as any);
    }
}
