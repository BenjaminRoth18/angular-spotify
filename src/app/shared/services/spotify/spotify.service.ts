import { HttpHeaders } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { PlaylistDto, PlaylistItemDto, Track } from 'src/app/models/track.model';

import { StorageService } from '../storage/storage.service';

@Injectable()
export class SpotifyService {
    tokenEndpoint = '/api/token';
    tokenParams = 'grant_type=client_credentials';
    playlistEndpoint = '/playlists';

    constructor(private storageService: StorageService) {}

    createTokenHeaders() {
        const headerOptions = {
            headers: new HttpHeaders({
                'Content-Type': 'application/x-www-form-urlencoded',
                Authorization:
                    'Basic OTk2MDgwOTM3ZWJiNDU5NGEwOTc5MTQ2YzljMGMxMjE6MGJkYTNjZmQyMTNjNDYyMmJjNmM1NjI1ODY1NjhlYzg='
            })
        };

        return headerOptions;
    }

    createAuthHeaders() {
        const token = this.storageService.getToken();

        const headerOptions = {
            headers: new HttpHeaders({
                Authorization: `Bearer ${token}`
            })
        };

        return headerOptions;
    }

    createPlaylist(playlistData: PlaylistDto): Track[] {
        const tracks = playlistData.tracks.items;

        return tracks.reduce((list: Track[], item: PlaylistItemDto) => {
            const trackItem = {} as Track;
            trackItem['artist'] = item.track.artists[0].name;
            trackItem['track'] = item.track.name;
            trackItem['imageUrl'] = item.track.album.images[0].url;
            list.push(trackItem);
            return list;
        }, []);
    }
}
