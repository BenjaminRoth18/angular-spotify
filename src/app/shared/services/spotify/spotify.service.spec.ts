import { playlistDtoMock } from 'src/app/models/track.mock';

import { MockStorageService } from '../storage/storage.service.mock';
import { SpotifyService } from './spotify.service';

describe('SpotifyService', () => {
    let spotifyService: SpotifyService;
    let storageService: MockStorageService;

    beforeEach(() => {
        storageService = new MockStorageService();
        spotifyService = new SpotifyService(storageService);
    });

    it('should create token headers', () => {
        const headers = spotifyService.createTokenHeaders();
        expect(headers.headers.get('Content-Type')).toEqual(
            'application/x-www-form-urlencoded'
        );
        expect(headers.headers.get('Authorization')).toEqual(
            'Basic OTk2MDgwOTM3ZWJiNDU5NGEwOTc5MTQ2YzljMGMxMjE6MGJkYTNjZmQyMTNjNDYyMmJjNmM1NjI1ODY1NjhlYzg='
        );
    });

    it('should create auth headers', () => {
        storageService.getToken = jest.fn().mockReturnValue('token');

        const headers = spotifyService.createAuthHeaders();

        expect(headers.headers.get('Authorization')).toEqual('Bearer token');
    });

    it('should create playlist snapshot', () => {
        expect(
            spotifyService.createPlaylist(playlistDtoMock)
        ).toMatchSnapshot();
    });
});
