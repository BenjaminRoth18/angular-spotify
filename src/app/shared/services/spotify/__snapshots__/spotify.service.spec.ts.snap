// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpotifyService should create playlist snapshot 1`] = `
Array [
  Object {
    "artist": "artist name",
    "imageUrl": "image url",
    "track": "track name",
  },
]
`;
